include/irr/asset/material_compiler/IR.h

```diff
@@ -0,0 +1,233 @@
+#ifndef __IRR_MATERIAL_COMPILER_IR_H_INCLUDED__
+#define __IRR_MATERIAL_COMPILER_IR_H_INCLUDED__
+
+#include <irr/core/IReferenceCounted.h>
+#include <irr/core/memory/refctd_dynamic_array.h>
+#include <irr/asset/ICPUImageView.h>
+#include <irr/asset/ICPUSampler.h>
+
+namespace irr {
+namespace asset {
+namespace material_compiler
+{
+
+class IR : public core::IReferenceCounted
+{
+public:
+    struct INode : public core::IReferenceCounted
+    {
+        virtual ~INode() = default;
+
+        enum E_SYMBOL
+        {
+            ES_MATERIAL,
+            ES_GEOM_MODIFIER,
+            ES_FRONT_SURFACE,
+            ES_BACK_SURFACE,
+            ES_EMISSION,
+            ES_BSDF,
+            ES_BSDF_COMBINER
+        };
+        enum E_PARAM_SOURCE
+        {
+            EPS_CONSTANT,
+            EPS_TEXTURE
+        };
+
+        struct STextureSource {
+            core::smart_refctd_ptr<ICPUImageView> image;
+            core::smart_refctd_ptr<ICPUSampler> sampler;
+            float scale;
+
+            bool operator==(const STextureSource& rhs) const { return image==rhs.image && sampler==rhs.sampler && scale==rhs.scale; }
+        };
+        //destructor of actually used variant member has to be called by hand!
+        template <typename type_of_const>
+        union UTextureOrConstant {
+            UTextureOrConstant() : texture{nullptr,nullptr,0.f} {}
+            ~UTextureOrConstant() {}
+
+            type_of_const constant;
+            STextureSource texture;
+        };
+        template <typename type_of_const>
+        struct SParameter
+        {
+            ~SParameter() {
+                if (source==EPS_TEXTURE)
+                    value.texture.~STextureSource();
+            }
+
+            SParameter<type_of_const>& operator=(const SParameter<type_of_const>& rhs)
+            {
+                source = rhs.source;
+                if (source == EPS_CONSTANT)
+                    value.constant = rhs.value.constant;
+                else
+                    value.texture = rhs.value.texture;
+            }
+
+            bool operator==(const SParameter<type_of_const>& rhs) const {
+                if (source!=rhs.source)
+                    return false;
+                switch (source)
+                {
+                case EPS_CONSTANT:
+                    return value.constant==value.constant;
+                case EPS_TEXTURE:
+                    return value.texture==value.texture;
+                default: return false;
+                }
+            }
+
+            E_PARAM_SOURCE source;
+            UTextureOrConstant<type_of_const> value;
+        };
+
+        using children_smart_array_t = core::smart_refctd_dynamic_array<core::smart_refctd_ptr<INode>>;
+
+        using color_t = core::vector3df_SIMD;
+
+        INode(E_SYMBOL s) : symbol(s) {}
+
+        children_smart_array_t children;
```

Comment:
if you want this to be performant and not completely kill the performance, then children need to be an offset+count (or multiple offsets) into a shared array of nodes for the whole IR

having a dynamically allocated mini-array will eat loads of memory (because of minimum allocation sizes) and completely choke out the `malloc` (the more allocations you have, the more time the next `malloc/free` takes, simply because there are more memory blocks to traverse upon either `malloc` or `free`)

Reply:
yea, i even wanted to put all nodes in one preallocated memory with placement new and linear addr allctr, but do it in dirty way first to not mess up